### PR TITLE
Remove LocalDateTimeArray, introduce TemporalArray

### LINKED ISSUE #474: Implement support for converting Date{32,64}Array,TimestampArray to/from arrow
Currently inferring from Arrow -> Vortex schema will throw if we attempt to use a column with one of these types. We also can't convert into these types currently from Canonical.

### PR DESCRIPTION
**Pull request by a10y**
Closes #474 

This PR removes the old `LocalDateTimeArray`, and replaces it with a more expansive `TemporalArray`.

`TemporalArray` encapsulates all temporal datatypes from arrow.

New extension DTypes are defined for each of the relevant Arrow types:

* `arrow.time32`
* `arrow.time64`
* `arrow.date32`
* `arrow.timestamp`
* `arrow.duration` (**still working on this**)
* `arrow.interval` (**still working on this**)

Each has a corresponding public constructor with docstrings. There are conformance tests for all of these variants to ensure the preconditions are being met around ptype and time_units allowed for each kind.

The compressor for DateTimeParts will only attempt to compress `arrow.timestamp`, as before. Unlike before, timestamps can carry timezones, which is serialized as part of the extension metadata. See the `TemporalMetadata` type in `temporal.rs`.

## TPC-H

After this change, we can successfully update the schema for our TPC-H datafusion benchmark to use `DataType::Date32` for all date columns, replacing `Utf8`. Not having to do the casting/parsing at execution time seems to have provided a tidy speedup.

### CODE REVIEW

**Comment by a10y** [comment]
I've update the PR per conversation with @gatesn and @robert3005 

* `arrow.time32` / `arrow.time64` -> `vortex.time`
* `arrow.date32` / `arrow.date64` -> `vortex.date`
* `arrow.timestamp` -> `vortex.timestamp`

The time_unit is stored in the ExtMetadata for all 3 types.

Note: we need to store the time_unit for Date because we need to support conversions between Arrow/Vortex schemas without retrieving any data. So even though the width of the backing primitive array is technically enough to infer the time_unit, we want to ensure that we can infer the schema without access to the data

File: encodings/datetime-parts/src/lib.rs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+extern crate core;
```
Comment (robert3005) [inline encodings/datetime-parts/src/lib.rs]: This should be gone, don't understand why intellij does this

Comment (a10y) [reply encodings/datetime-parts/src/lib.rs]: done

File: vortex-array/src/array/datetime/mod.rs [hunk 2]
```diff
@@ -28,33 +30,59 @@ pub enum TimeUnit {
...
-    }
+    D,
 }
 
 impl Display for TimeUnit {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         match self {
             Self::Ns => write!(f, "ns"),
-            Self::Us => write!(f, "us"),
+            Self::Us => write!(f, "µs"),
```
Comment (a10y) [inline vortex-array/src/array/datetime/mod.rs:40]: image

File: vortex-array/src/arrow/dtype.rs [hunk 3]
```diff
@@ -68,14 +68,10 @@ impl FromArrowType<&Field> for DType {
...
             DataType::Binary | DataType::LargeBinary => Binary(nullability),
-            DataType::Timestamp(time_unit, tz) => match tz {
-                None => Extension(LocalDateTimeArray::ext_dtype(time_unit.into()), nullability),
-                Some(_) => unimplemented!("Timezone not yet supported"),
-            },
-            // DataType::Date32 => localdate(IntWidth::_32, nullability),
-            // DataType::Date64 => localdate(IntWidth::_64, nullability),
-            // DataType::Time32(u) => localtime(u.into(), IntWidth::_32, nullability),
-            // DataType::Time64(u) => localtime(u.into(), IntWidth::_64, nullability),
+            DataType::Time32(_) | DataType::Time64(_) | DataType::Timestamp(..) => Extension(
```
Comment (robert3005) [inline vortex-array/src/arrow/dtype.rs]: Should this handle Date32/64?

Comment (a10y) [reply vortex-array/src/arrow/dtype.rs]: done

**Review by robert3005 — APPROVED** [review]
I wish there was a better way of encoding timezones in metadata but can't really come up with anything that doesn't involve including timezone database with us

**Comment by a10y** [comment]
Agreed

**Pull request merged**